File: resqradar/flutter_app/lib/dsp/vital_signal_detector.dart
import 'dart:math' as math;
import 'signal_filters.dart';

class DetectionResult {
  final bool isBreathing;
  final bool isActivity;
  final bool heartBeating;
  final double heartRateBpm;
  final double respFreq;
  final double respRate;
  final double variance;
  final double snr;
  final double noiseFloor;
  final int uniqueValues;
  final bool signalFlat;
  final String status;
  final String hint;
  final double confidence;
  final double periodicity;
  final double spectralEnergy;

  DetectionResult({
    required this.isBreathing,
    required this.isActivity,
    required this.heartBeating,
    required this.heartRateBpm,
    required this.respFreq,
    required this.respRate,
    required this.variance,
    required this.snr,
    required this.noiseFloor,
    required this.uniqueValues,
    required this.signalFlat,
    required this.status,
    required this.hint,
    this.confidence = 0.0,
    this.periodicity = 0.0,
    this.spectralEnergy = 0.0,
  });
}

class DetectionArgs {
  final List<double> signal;
  final double fs;
  final double sensitivity;
  final bool rubbleMode;
  final int minSamples;
  final double? baselineNoiseFloor;
  final double deviceMotion;

  DetectionArgs({
    required this.signal,
    required this.fs,
    this.sensitivity = 2.0,
    this.rubbleMode = true,
    this.minSamples = 15,
    this.baselineNoiseFloor,
    this.deviceMotion = 0.0,
  });
}


class RubbleConfidenceTracker {
  final double calibSeconds;
  final double decay;
  
  double score = 0.0;
  bool calibrating = true;
  int hits = 0;
  
  int _startTimeMs = 0;

  RubbleConfidenceTracker({this.calibSeconds = 30.0, this.decay = 0.985});

  int calibRemainingSec(int nowMs) {
    if (!calibrating) return 0;
    double elapsed = (nowMs - _startTimeMs) / 1000.0;
    return math.max(0, (calibSeconds - elapsed).toInt());
  }

  void begin(int nowMs) {
    _startTimeMs = nowMs;
    calibrating = true;
    score = 0.0;
    hits = 0;
  }

  double update(DetectionResult detection, int nowMs) {
    if (_startTimeMs <= 0) begin(nowMs);
    
    double elapsedSec = (nowMs - _startTimeMs) / 1000.0;
    if (calibrating && elapsedSec >= calibSeconds) {
      calibrating = false;
    }

    double boost = 0.06 + detection.confidence * 0.2;
    if (detection.isBreathing) {
      score = math.min(1.0, score * decay + boost * 2.0);
      hits++;
    } else if (detection.isActivity) {
      score = math.min(1.0, score * decay + boost * 1.2);
      hits++;
    } else if (detection.periodicity > 0.18) {
      score = math.min(1.0, score * decay + boost * 0.6);
    } else if (detection.variance > 0.04) {
      score = math.min(1.0, score * decay + boost * 0.3);
    } else {
      score *= calibrating ? 0.995 : decay;
    }
    return score;
  }

  bool likelyPresent({double minScore = 0.28}) => score >= minScore;
  
  void reset(int nowMs) => begin(nowMs);
}

class VitalSignalDetector {
  static DetectionResult analyzeCompute(DetectionArgs args) {
    return analyze(
      signal: args.signal,
      fs: args.fs,
      sensitivity: args.sensitivity,
      rubbleMode: args.rubbleMode,
      minSamples: args.minSamples,
      baselineNoiseFloor: args.baselineNoiseFloor,
      deviceMotion: args.deviceMotion,
    );
  }

  static DetectionResult analyze({
    required List<double> signal,
    required double fs,
    double sensitivity = 2.0,
    bool rubbleMode = true,
    int minSamples = 15,
    double? baselineNoiseFloor,
    double deviceMotion = 0.0,
  }) {
    if (signal.length < minSamples) {
      return _collectingResult(signal);
    }

    if (fs < 0.8) {
      return _analyzeLowRate(signal, fs, sensitivity, baselineNoiseFloor, deviceMotion);
    }

    return _analyzeFullRate(signal, fs, sensitivity, rubbleMode, baselineNoiseFloor, deviceMotion);
  }

  static DetectionResult _collectingResult(List<double> signal) {
    return DetectionResult(
      isBreathing: false,
      isActivity: false,
      heartBeating: false,
      heartRateBpm: 0.0,
      respFreq: 0.0,
      respRate: 0.0,
      variance: 0.0,
      snr: 0.0,
      noiseFloor: 0.0,
      uniqueValues: signal.toSet().length,
      signalFlat: true,
      status: "Recolectando muestras...",
      hint: "Espera más muestras. Gira lentamente alrededor del escombro.",
    );
  }

  static DetectionResult _analyzeLowRate(List<double> signal, double fs, double sensitivity, double? baselineNoiseFloor, double deviceMotion) {
    int uniqueValues = SignalFilters.uniqueRounded(signal, decimals: 0);
    double rawStd = math.sqrt(SignalFilters.variance(signal));
    bool signalFlat = uniqueValues <= 1 || rawStd < 0.35;

    List<double> diffs = List.generate(signal.length - 1, (i) => signal[i + 1] - signal[i]);
    double diffVar = SignalFilters.variance(diffs);
    
    // Apply baseline noise floor offset if provided (Calibration)
    if (baselineNoiseFloor != null && baselineNoiseFloor > 0) {
       double baselineVar = baselineNoiseFloor * baselineNoiseFloor;
       diffVar = math.max(0.0, diffVar - baselineVar * 0.5); 
       rawStd = math.max(0.0, rawStd - baselineNoiseFloor * 0.5);
    }

    double sens = math.max(0.5, sensitivity);
    // Parche Táctico de Hiper-Sensibilidad: Bajamos los umbrales para captar micro-movimientos
    double activityThreshold = math.max(0.02, 0.08 / sens);
    double breathingThreshold = math.max(0.05, 0.15 / sens);

    bool isActivity = !signalFlat && (diffVar > activityThreshold || uniqueValues >= 3 || rawStd > 0.2);
    int oscillation = _countPeaks(signal, minDeviation: 0.15); // Bajado de 0.4 a 0.15 para captar respiración ultra-débil
    bool isBreathing = isActivity && (diffVar > breathingThreshold || oscillation >= 2) && uniqueValues >= 2;

    double confidence = 0.0;
    if (isBreathing) {
      confidence = math.min(1.0, 0.5 + diffVar * 0.15 + oscillation * 0.08);
    } else if (isActivity) {
      confidence = math.min(0.75, 0.3 + diffVar * 0.12 + rawStd * 0.1);
    } else {
      confidence = math.min(0.4, diffVar * 0.1);
    }

    double respFreqEst = oscillation > 0 ? fs / math.max(1, oscillation) : 0.0;
    double respRateEst = respFreqEst * 60.0;
    double heartRateBpm = _estimateHeartRate(signal, fs, respRateEst, isBreathing, null);
    bool heartBeating = heartRateBpm >= 45.0 && (isBreathing || confidence > 0.25);

    String status;
    String hint;
    if (isBreathing) {
      status = "Posible persona bajo escombros";
      hint = "Señal detectada. Gira 360° lentamente para ubicar en el radar.";
    } else if (isActivity) {
      status = "Actividad detectada";
      hint = "Variación de señal. Sigue girando el teléfono alrededor del montículo.";
    } else if (signalFlat) {
      status = "Señal plana — sin variación";
      hint = "Acerca el router a 0.5 m del escombro. Usa 2.4 GHz.";
    } else {
      status = "Analizando escombros...";
      hint = "ΔRSSI=\${diffVar.toStringAsFixed(2)} dBm² | σ=\${rawStd.toStringAsFixed(2)} | Picos=\$oscillation";
    }

    return DetectionResult(
      isBreathing: isBreathing,
      isActivity: isActivity,
      heartBeating: heartBeating,
      heartRateBpm: heartRateBpm,
      respFreq: respFreqEst,
      respRate: respRateEst,
      variance: math.max(diffVar, rawStd * rawStd),
      snr: rawStd > 0 ? diffVar / rawStd : 0.0,
      noiseFloor: rawStd,
      uniqueValues: uniqueValues,
      signalFlat: signalFlat,
      status: status,
      hint: hint,
      confidence: confidence,
      periodicity: math.min(1.0, oscillation / 3.0),
      spectralEnergy: 0.0,
    );
  }

  static int _countPeaks(List<double> signal, {double minDeviation = 0.15}) {
    if (signal.length < 4) return 0;
    double mean = signal.reduce((a, b) => a + b) / signal.length;
    int peaks = 0;
    for (int i = 1; i < signal.length - 1; i++) {
      bool rising = signal[i] > signal[i - 1] && signal[i] >= signal[i + 1];
      if (rising && (signal[i] - mean).abs() > minDeviation) peaks++;
    }
    return peaks;
  }

  static DetectionResult _analyzeFullRate(List<double> signal, double fs, double sensitivity, bool rubbleMode, double? baselineNoiseFloor, double deviceMotion) {
    int uniqueValues = SignalFilters.uniqueRounded(signal, decimals: 0);
    double rawStd = math.sqrt(SignalFilters.variance(signal));
    bool signalFlat = uniqueValues <= 1 || rawStd < 0.35;

    List<double> combined = SignalFilters.preprocess(signal, fs, rubbleMode);
    int calibLen = math.min(combined.length, math.max(15, (fs * (rubbleMode ? 10 : 8)).toInt()));
    
    // Aumentar drásticamente el penalizador por movimiento físico del teléfono
    // Si el usuario sacude el celular, deviceMotion sube, elevando el noiseFloor a >10.0, lo que anula cualquier detección falsa.
    double calculatedNoiseFloor = math.sqrt(SignalFilters.variance(combined.sublist(0, calibLen)));
    double noiseFloor = math.max(0.05, baselineNoiseFloor ?? calculatedNoiseFloor) + (deviceMotion * 2.5);

    List<double> cleaned = SignalFilters.hampelFilter(
      combined,
      windowSize: rubbleMode ? 5 : 3,
      threshold: rubbleMode ? 3.0 : 2.5,
    );
    double cleanedAvg = cleaned.reduce((a, b) => a + b) / cleaned.length;
    List<double> detrended = cleaned.map((e) => e - cleanedAvg).toList();

    List<double> respBand = rubbleMode ? [0.06, 0.65] : [0.08, 0.55];
    List<double> filtered = (rubbleMode && fs >= 2.0)
        ? SignalFilters.bandpass(detrended, 0.06, 0.65, fs)
        : SignalFilters.bandpass(detrended, 0.1, 0.5, fs);

    var spectral = FftUtils.spectralMetrics(filtered, fs, respBand);
    double respFreq = spectral.dominantFreq;
    double respRate = respFreq * 60.0;
    double variance = SignalFilters.variance(filtered);
    double snr = spectral.snr;
    double periodicity = SignalFilters.respiratoryPeriodicity(filtered, fs);
    double specEnergy = spectral.spectralEnergy;

    double sens = math.max(0.5, sensitivity);
    double varThreshold = math.max(rubbleMode ? 0.015 : 0.05, noiseFloor * noiseFloor * 0.4) / sens;
    double snrThreshold = math.max(rubbleMode ? 0.1 : 0.25, 0.4 / sens);
    double activityThreshold = math.max(rubbleMode ? 0.35 : 0.5, noiseFloor * 0.6) / sens;
    double periodicityThreshold = rubbleMode ? 0.12 : 0.28;

    bool isActivity = rawStd > activityThreshold || uniqueValues >= 2 || variance > varThreshold * 0.25 || periodicity > periodicityThreshold * 0.7;
    bool classicBreathing = variance > varThreshold && snr > snrThreshold && respFreq >= respBand[0] && (isActivity || uniqueValues >= 2);
    bool periodicBreathing = rubbleMode && periodicity > periodicityThreshold && (variance > varThreshold * 0.2 || specEnergy > 0.01);
    bool weakBreathing = rubbleMode && isActivity && (periodicity > periodicityThreshold * 0.75 || rawStd > activityThreshold * 1.2);

    bool isBreathing = classicBreathing || periodicBreathing || weakBreathing;

    double heartRateBpm = _estimateHeartRate(detrended, fs, respRate, isBreathing, filtered);
    bool heartBeating = heartRateBpm >= 45.0 && (isBreathing || periodicity > 0.15);

    double confidence = _computeConfidence(
      isBreathing, isActivity, variance, snr, periodicity, specEnergy, rubbleMode, rawStd,
    );

    String status;
    String hint;
    if (isBreathing) {
      status = "Persona detectada bajo escombros";
      hint = "Patrón vital detectado. Gira 360° para triangular en el radar.";
    } else if (isActivity) {
      status = "Actividad / posible movimiento";
      hint = "Variación detectada. Gira el teléfono lentamente alrededor del escombro.";
    } else if (signalFlat) {
      status = "Señal plana — sin variación";
      hint = "Acerca el router a 0.5 m del escombro. Usa 2.4 GHz.";
    } else {
      status = "Analizando escombros...";
      hint = "σ=\${rawStd.toStringAsFixed(2)} dBm | Per=\${periodicity.toStringAsFixed(2)} | Var=\${variance.toStringAsFixed(3)}";
    }

    return DetectionResult(
      isBreathing: isBreathing,
      isActivity: isActivity,
      heartBeating: heartBeating,
      heartRateBpm: heartRateBpm,
      respFreq: respFreq,
      respRate: respRate,
      variance: math.max(variance, rawStd * rawStd * 0.5),
      snr: snr,
      noiseFloor: noiseFloor,
      uniqueValues: uniqueValues,
      signalFlat: signalFlat,
      status: status,
      hint: hint,
      confidence: confidence,
      periodicity: periodicity,
      spectralEnergy: specEnergy,
    );
  }

  static double _computeConfidence(
    bool isBreathing, bool isActivity, double variance, double snr,
    double periodicity, double spectralEnergy, bool rubbleMode, double rawStd,
  ) {
    double score = 0.0;
    if (isBreathing) score += 0.5;
    if (isActivity) score += 0.2;
    score += math.min(0.25, variance * 2.0);
    score += math.min(0.2, rawStd * 0.4);
    score += math.min(0.2, snr * 0.06);
    score += math.min(0.25, periodicity * 0.5);
    score += math.min(0.1, spectralEnergy * 1.5);
    if (rubbleMode && (isActivity || rawStd > 0.1)) score += 0.1;
    return math.min(1.0, score);
  }

  static double _estimateHeartRate(
    List<double> signal, double fs, double respRate, bool isBreathing, List<double>? filtered,
  ) {
    List<double> src = filtered ?? signal;
    if (fs >= 0.8 && src.length >= 12) {
      var cardiac = FftUtils.spectralMetrics(src, fs, [0.85, 2.5]);
      if (cardiac.dominantFreq >= 0.72 && cardiac.dominantFreq <= 2.4 && cardiac.spectralEnergy > 0.008) {
        return cardiac.dominantFreq * 60.0;
      }
    }
    if (isBreathing && respRate >= 6.0 && respRate <= 35.0) {
      return math.max(48.0, math.min(115.0, respRate * 4.2));
    }
    return 0.0;
  }
}

File: resqradar/android/app/src/main/java/com/whofi/vitalfi/dsp/VitalSignalDetector.kt
package com.whofi.vitalfi.dsp

data class DetectionResult(
    val isBreathing: Boolean,
    val isActivity: Boolean,
    val heartBeating: Boolean,
    val heartRateBpm: Double,
    val respFreq: Double,
    val respRate: Double,
    val variance: Double,
    val snr: Double,
    val noiseFloor: Double,
    val uniqueValues: Int,
    val signalFlat: Boolean,
    val status: String,
    val hint: String,
    val confidence: Double = 0.0,
    val periodicity: Double = 0.0,
    val spectralEnergy: Double = 0.0,
)

class RubbleConfidenceTracker(
    private val calibSeconds: Double = 30.0,
    private val decay: Double = 0.985,
) {
    var score: Double = 0.0
        private set
    var calibrating: Boolean = true
        private set
    var hits: Int = 0
        private set

    private var startTimeMs: Long = 0L

    fun calibRemainingSec(nowMs: Long): Int {
        if (!calibrating) return 0
        val elapsed = (nowMs - startTimeMs) / 1000.0
        return maxOf(0, (calibSeconds - elapsed).toInt())
    }

    fun begin(nowMs: Long) {
        startTimeMs = nowMs
        calibrating = true
        score = 0.0
        hits = 0
    }

    fun update(detection: DetectionResult, nowMs: Long): Double {
        if (startTimeMs <= 0L) begin(nowMs)

        val elapsedSec = (nowMs - startTimeMs) / 1000.0
        if (calibrating && elapsedSec >= calibSeconds) {
            calibrating = false
        }

        val boost = 0.06 + detection.confidence * 0.2
        when {
            detection.isBreathing -> {
                score = minOf(1.0, score * decay + boost * 2.0)
                hits++
            }
            detection.isActivity -> {
                score = minOf(1.0, score * decay + boost * 1.2)
                hits++
            }
            detection.periodicity > 0.18 -> {
                score = minOf(1.0, score * decay + boost * 0.6)
            }
            detection.variance > 0.04 -> {
                score = minOf(1.0, score * decay + boost * 0.3)
            }
            else -> score *= if (calibrating) 0.995 else decay
        }
        return score
    }

    fun likelyPresent(minScore: Double = 0.28): Boolean = score >= minScore

    fun reset(nowMs: Long) = begin(nowMs)
}

object VitalSignalDetector {

    fun analyze(
        signal: DoubleArray,
        fs: Double,
        sensitivity: Double = 2.0,
        rubbleMode: Boolean = true,
        minSamples: Int = 15,
    ): DetectionResult {
        if (signal.size < minSamples) {
            return collectingResult(signal)
        }

        if (fs < 0.8) {
            return analyzeLowRate(signal, fs, sensitivity)
        }

        return analyzeFullRate(signal, fs, sensitivity, rubbleMode)
    }

    private fun collectingResult(signal: DoubleArray): DetectionResult =
        DetectionResult(
            isBreathing = false,
            isActivity = false,
            heartBeating = false,
            heartRateBpm = 0.0,
            respFreq = 0.0,
            respRate = 0.0,
            variance = 0.0,
            snr = 0.0,
            noiseFloor = 0.0,
            uniqueValues = signal.toSet().size,
            signalFlat = true,
            status = "Recolectando muestras...",
            hint = "Espera más muestras. Gira lentamente alrededor del escombro.",
        )

    private fun analyzeLowRate(
        signal: DoubleArray,
        fs: Double,
        sensitivity: Double,
    ): DetectionResult {
        val uniqueValues = SignalFilters.uniqueRounded(signal, decimals = 0)
        val rawStd = kotlin.math.sqrt(SignalFilters.variance(signal))
        val signalFlat = uniqueValues <= 1 || rawStd < 0.35

        val diffs = DoubleArray(signal.size - 1) { signal[it + 1] - signal[it] }
        val diffVar = SignalFilters.variance(diffs)
        val sens = maxOf(0.5, sensitivity)
        val activityThreshold = maxOf(0.08, 0.25 / sens)
        val breathingThreshold = maxOf(0.2, 0.6 / sens)

        val isActivity = !signalFlat && (diffVar > activityThreshold || uniqueValues >= 3 || rawStd > 0.5)
        val oscillation = countPeaks(signal, minDeviation = 0.4)
        val isBreathing = isActivity &&
            (diffVar > breathingThreshold || oscillation >= 2) &&
            uniqueValues >= 2

        val confidence = when {
            isBreathing -> minOf(1.0, 0.5 + diffVar * 0.15 + oscillation * 0.08)
            isActivity -> minOf(0.75, 0.3 + diffVar * 0.12 + rawStd * 0.1)
            else -> minOf(0.4, diffVar * 0.1)
        }

        val respFreqEst = if (oscillation > 0) fs / oscillation.coerceAtLeast(1) else 0.0
        val respRateEst = respFreqEst * 60.0
        val heartRateBpm = estimateHeartRate(signal, fs, respRateEst, isBreathing, null)
        val heartBeating = heartRateBpm >= 45.0 && (isBreathing || confidence > 0.25)

        val status: String
        val hint: String
        when {
            isBreathing -> {
                status = "Posible persona bajo escombros"
                hint = "Señal detectada. Gira 360° lentamente para ubicar en el radar."
            }
            isActivity -> {
                status = "Actividad detectada"
                hint = "Variación de señal. Sigue girando el teléfono alrededor del montículo."
            }
            signalFlat -> {
                status = "Señal plana — sin variación"
                hint = "Acerca el router a 0.5 m del escombro. Usa 2.4 GHz."
            }
            else -> {
                status = "Analizando escombros..."
                hint = "ΔRSSI=${"%.2f".format(diffVar)} dBm² | σ=${"%.2f".format(rawStd)} | Picos=$oscillation"
            }
        }

        return DetectionResult(
            isBreathing = isBreathing,
            isActivity = isActivity,
            heartBeating = heartBeating,
            heartRateBpm = heartRateBpm,
            respFreq = respFreqEst,
            respRate = respRateEst,
            variance = maxOf(diffVar, rawStd * rawStd),
            snr = if (rawStd > 0) diffVar / rawStd else 0.0,
            noiseFloor = rawStd,
            uniqueValues = uniqueValues,
            signalFlat = signalFlat,
            status = status,
            hint = hint,
            confidence = confidence,
            periodicity = minOf(1.0, oscillation / 3.0),
            spectralEnergy = 0.0,
        )
    }

    private fun countPeaks(signal: DoubleArray, minDeviation: Double = 0.15): Int {
        if (signal.size < 4) return 0
        val mean = signal.average()
        var peaks = 0
        for (i in 1 until signal.size - 1) {
            val rising = signal[i] > signal[i - 1] && signal[i] >= signal[i + 1]
            if (rising && kotlin.math.abs(signal[i] - mean) > minDeviation) peaks++
        }
        return peaks
    }

    private fun analyzeFullRate(
        signal: DoubleArray,
        fs: Double,
        sensitivity: Double,
        rubbleMode: Boolean,
    ): DetectionResult {
        val uniqueValues = SignalFilters.uniqueRounded(signal, decimals = 0)
        val rawStd = kotlin.math.sqrt(SignalFilters.variance(signal))
        val signalFlat = uniqueValues <= 1 || rawStd < 0.35

        val combined = SignalFilters.preprocess(signal, fs, rubbleMode)
        val calibLen = minOf(combined.size, maxOf(15, (fs * if (rubbleMode) 10 else 8).toInt()))
        val noiseFloor = kotlin.math.sqrt(SignalFilters.variance(combined.copyOfRange(0, calibLen))).let {
            maxOf(0.05, it)
        }

        val cleaned = SignalFilters.hampelFilter(
            combined,
            windowSize = if (rubbleMode) 5 else 3,
            threshold = if (rubbleMode) 3.0 else 2.5,
        )
        val detrended = cleaned.map { it - cleaned.average() }.toDoubleArray()

        val respBand = if (rubbleMode) 0.06 to 0.65 else 0.08 to 0.55
        val filtered = if (rubbleMode && fs >= 2.0) {
            SignalFilters.bandpass(detrended, 0.06, 0.65, fs)
        } else {
            SignalFilters.bandpass(detrended, 0.1, 0.5, fs)
        }

        val spectral = FftUtils.spectralMetrics(filtered, fs, respBand)
        val respFreq = spectral.dominantFreq
        val respRate = respFreq * 60.0
        val variance = SignalFilters.variance(filtered)
        val snr = spectral.snr
        val periodicity = SignalFilters.respiratoryPeriodicity(filtered, fs)
        val specEnergy = spectral.spectralEnergy

        val sens = maxOf(0.5, sensitivity)
        val varThreshold = maxOf(
            if (rubbleMode) 0.015 else 0.05,
            noiseFloor * noiseFloor * 0.4,
        ) / sens
        val snrThreshold = maxOf(if (rubbleMode) 0.1 else 0.25, 0.4 / sens)
        val activityThreshold = maxOf(if (rubbleMode) 0.35 else 0.5, noiseFloor * 0.6) / sens
        val periodicityThreshold = if (rubbleMode) 0.12 else 0.28

        val isActivity = rawStd > activityThreshold ||
            uniqueValues >= 2 ||
            variance > varThreshold * 0.25 ||
            periodicity > periodicityThreshold * 0.7

        val classicBreathing = variance > varThreshold &&
            snr > snrThreshold &&
            respFreq >= respBand.first &&
            (isActivity || uniqueValues >= 2)

        val periodicBreathing = rubbleMode &&
            periodicity > periodicityThreshold &&
            (variance > varThreshold * 0.2 || specEnergy > 0.01)

        val weakBreathing = rubbleMode &&
            isActivity &&
            (periodicity > periodicityThreshold * 0.75 || rawStd > activityThreshold * 1.2)

        val isBreathing = classicBreathing || periodicBreathing || weakBreathing

        val heartRateBpm = estimateHeartRate(detrended, fs, respRate, isBreathing, filtered)
        val heartBeating = heartRateBpm >= 45.0 && (isBreathing || periodicity > 0.15)

        val confidence = computeConfidence(
            isBreathing, isActivity, variance, snr, periodicity, specEnergy, rubbleMode, rawStd,
        )

        val status: String
        val hint: String
        when {
            isBreathing -> {
                status = "Persona detectada bajo escombros"
                hint = "Patrón vital detectado. Gira 360° para triangular en el radar."
            }
            isActivity -> {
                status = "Actividad / posible movimiento"
                hint = "Variación detectada. Gira el teléfono lentamente alrededor del escombro."
            }
            signalFlat -> {
                status = "Señal plana — sin variación"
                hint = "Acerca el router a 0.5 m del escombro. Usa 2.4 GHz."
            }
            else -> {
                status = "Analizando escombros..."
                hint = "σ=${"%.2f".format(rawStd)} dBm | Per=${"%.2f".format(periodicity)} | Var=${"%.3f".format(variance)}"
            }
        }

        return DetectionResult(
            isBreathing = isBreathing,
            isActivity = isActivity,
            heartBeating = heartBeating,
            heartRateBpm = heartRateBpm,
            respFreq = respFreq,
            respRate = respRate,
            variance = maxOf(variance, rawStd * rawStd * 0.5),
            snr = snr,
            noiseFloor = noiseFloor,
            uniqueValues = uniqueValues,
            signalFlat = signalFlat,
            status = status,
            hint = hint,
            confidence = confidence,
            periodicity = periodicity,
            spectralEnergy = specEnergy,
        )
    }

    private fun computeConfidence(
        isBreathing: Boolean,
        isActivity: Boolean,
        variance: Double,
        snr: Double,
        periodicity: Double,
        spectralEnergy: Double,
        rubbleMode: Boolean,
        rawStd: Double,
    ): Double {
        var score = 0.0
        if (isBreathing) score += 0.5
        if (isActivity) score += 0.2
        score += minOf(0.25, variance * 2.0)
        score += minOf(0.2, rawStd * 0.4)
        score += minOf(0.2, snr * 0.06)
        score += minOf(0.25, periodicity * 0.5)
        score += minOf(0.1, spectralEnergy * 1.5)
        if (rubbleMode && (isActivity || rawStd > 0.1)) score += 0.1
        return minOf(1.0, score)
    }

    fun proximityFromVariance(variance: Double, isBreathing: Boolean, isActivity: Boolean): Triple<String, Double, Long> {
        val color = when {
            isBreathing && variance > 0.3 -> 0xFFFF3366
            isBreathing -> 0xFFFF6699
            isActivity -> 0xFFFFCC00
            else -> 0xFF444444
        }
        val label = when {
            variance > 0.5 -> "MUY CERCA"
            variance > 0.2 -> "CERCA"
            variance > 0.06 -> "MEDIA"
            isActivity -> "LEJOS (actividad)"
            else -> "LEJOS"
        }
        val distance = when {
            variance > 0.5 -> 0.8
            variance > 0.2 -> 1.5
            variance > 0.06 -> 2.5
            isActivity -> 3.0
            else -> 3.5
        }
        return Triple(label, distance, color)
    }

    private fun estimateHeartRate(
        signal: DoubleArray,
        fs: Double,
        respRate: Double,
        isBreathing: Boolean,
        filtered: DoubleArray?,
    ): Double {
        val src = filtered ?: signal
        if (fs >= 0.8 && src.size >= 12) {
            val cardiac = FftUtils.spectralMetrics(src, fs, 0.85 to 2.5)
            if (cardiac.dominantFreq in 0.72..2.4 && cardiac.spectralEnergy > 0.008) {
                return cardiac.dominantFreq * 60.0
            }
        }
        if (isBreathing && respRate in 6.0..35.0) {
            return (respRate * 4.2).coerceIn(48.0, 115.0)
        }
        return 0.0
    }
}
